Fix Issue 21050 - __traits(getOverloads) for templates returns incorr…

rmanthorpe opened the pull request:
…ect symbol for the first overload

UplinkCoder commented:
There is no need for new AST node to support the desired effect as far as I can see.

rmanthorpe commented:
Yeah I didn't really like this solution. What's the alternative approach? I could try to clone the `TemplateDeclaration` and null out `overnext`.

UplinkCoder commented:
> What's the alternative approach? I could try to clone the TemplateDeclaration and null out overnext.

Sounds much better.

BorisCarvajal commented:
> > What's the alternative approach? I could try to clone the TemplateDeclaration and null out overnext.
> 
> Sounds much better.

But that will create two different instances for the same template arguments.

For example this code should print "instantiated" only one time.
```
struct S {
    static void foo(T)(int i) { pragma(msg, "instantiated"); }
    static void foo(T)(string s) { }
}

alias foo0 = __traits(getOverloads, S, "foo", true)[0];
alias bar0 = foo0!long;  // instantiated here
void main()
{
    S.foo!long(0);  // already instantiated
}
```

UplinkCoder commented:
@BorisCarvajal not necessarily.

But the concerns are valid, we'll have to see how it turns out, eventually a new flag like `cloned` has to be introduced for TemplateDeclaration.

rmanthorpe commented:
This may sound naive but... What if I `syntaxCopy`, run semantic, then copy the `instances` field so they share the hash table of instantiations.

UplinkCoder commented:
@rmanthorpe I have no idea if that would work, my assumption is that it violated some invariant, but try it.

rmanthorpe commented:
I'll give it a go. How problematic is it if templates get instantiated more than once?

UplinkCoder commented:
@rmanthorpe I can't say.
Maybe it'll do nothing bad, in the worst case it's silently disastrous.

rmanthorpe commented:
This is at least a problem that a new AST node doesn't have, though it has others.

rmanthorpe committed:
Fix Issue 21050 - __traits(getOverloads) for templates returns incorrect symbol for the first overload

Template declarations are nodes in a linked list of overloads.
Returning a template declaration from __traits(getOverloads)
returned the tail of that list of overloads. Instead we must
take a copy of the template declaration and remove references
to overnext and overroot so that it is no longer part of the
overload set.

rmanthorpe commented:
As far as I can work out any form of copying, even doing a `memcpy` and then nulling out `overnext`, results in duplicate template instances. The AST node approach does not. I've added a check for that to the test.

I think the only way to do this is going to involve having _something_ which has a pointer to the original `TemplateDeclaration`.  Either a new AST node, or adding a `cloned` field which contains the original.

NilsLankila commented:
The overhaed when building dparse is ~3MB. This goes in the wrong direction as currently we try to pay attention to this.
Alas there might be no other solution. I have #11461, that would compensate this change.

rmanthorpe commented:
@NilsLankila I'm not at all sure if this is the right way to go. I need to dig a bit deeper into where the multiple instantiations come from when I have time. A copy-without-overnext solution does seem like the best option if I can make it work.

NilsLankila commented:
okay, maybe turn it as a draft to make clear that it's not readey even if it passes the tests.
Also you have the PR for the "child" traits that's almost good, it just needs a changelog entry.

BorisCarvajal commented:
Creating an empty `TemplateDeclaration.instances` fixes the multiple instantiation problem.
Like this:
```
diff --git a/src/dmd/traits.d b/src/dmd/traits.d
index f6a7873a4..da6df2b6e 100644
--- a/src/dmd/traits.d
+++ b/src/dmd/traits.d
@@ -1081,7 +1081,28 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
                 if (!fd)
                 {
                     if(includeTemplates)
-                        exps.push(new DsymbolExp(Loc.initial, s, false));
+                    {
+                        if (auto td = s.isTemplateDeclaration())
+                        {
+                            if (td.instances is null)
+                            {
+                                // create an empty AA just to copy it
+                                scope ti = new TemplateInstance(Loc.initial, Id.empty, null);
+                                auto tib = TemplateInstanceBox(ti);
+                                td.instances[tib] = null;
+                                td.instances.clear();
+                            }
+                            auto oinstances = td.instances;
+                            td = cast(TemplateDeclaration)td.syntaxCopy(null);
+                            import core.stdc.string : memcpy;
+                            memcpy(cast(void*)td, cast(void*)s, __traits(classInstanceSize, TemplateDeclaration));
+                            // use the original instances table
+                            td.instances = oinstances;
+                            // make sure overnext is null to avoid overload set logic in the future
+                            td.overnext = null;
+                            exps.push(new DsymbolExp(Loc.initial, td, false));
+                        }
+                    }
                     return 0;
                 }
                 if (e.ident == Id.getVirtualFunctions && !fd.isVirtual())
```

rmanthorpe commented:
> Creating an empty TemplateDeclaration.instances fixes the multiple instantiation problem.

@BorisCarvajal Thank you! You're totally right I feel like an idiot! I tried this but it didn't occur to me that `instances` might be null :facepalm:

BorisCarvajal commented on src/dmd/traits.d:
```diff
@@ -1080,8 +1080,36 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
+                                    td.instances[tib] = null;
+                                    td.instances.clear();
+                                }
+                                auto oinstances = td.instances;
```
```suggestion
```
This line is not needed with `memcpy`. (I was originally using `dsymbolSemantic` instead and forgot to delete it)

BorisCarvajal commented on src/dmd/traits.d:
```diff
@@ -1080,8 +1080,36 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
+                                import core.stdc.string : memcpy;
+                                memcpy(cast(void*) td, cast(void*) s,
+                                        __traits(classInstanceSize, TemplateDeclaration));
+                                td.instances = oinstances;
```
```suggestion
```
ditto.

NilsLankila commented on src/dmd/traits.d:
```diff
@@ -1080,8 +1080,34 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
+                            // which shares the same `instances` cache but without
+                            // `overroot` and `overnext` set to avoid overload
+                            // behaviour in the result.
+                            if(td.overnext !is null)
```
there's a space missing between the keyword and the left parens.

rmanthorpe replied:
I'm not doing well at this. Fixed now thanks.
